Disallow dropping Iceberg schema that contains external files

jirassimok opened the pull request:
Based on https://github.com/trinodb/trino/pull/10146
Adds same logic for Iceberg

jirassimok committed:
Only delete non-empty directories when dropping Iceberg schemas

jirassimok committed:
Use Hive 3 Tempto config in Spark product test envs

findepi commented:
Shouldn't be merged in current shape -- https://github.com/trinodb/trino/pull/9740#issuecomment-955591970

jirassimok commented:
Fixed the tests; this should work properly now.

findepi commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/EnvSinglenodeSparkIceberg.java:
```diff
@@ -57,20 +59,25 @@ public EnvSinglenodeSparkIceberg(Standard standard, Hadoop hadoop, DockerFiles d
                         CONTAINER_PRESTO_ETC + "/catalog/iceberg.properties"));
-
-        builder.addContainer(createSpark())
-                .containerDependsOn("spark", HADOOP);
```
why removed? separate commit?

jirassimok replied:
Hm, I don't remember why, and it seems to work either way, so I'll put this back.

findepi commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/EnvSinglenodeSparkIceberg.java:
```diff
@@ -57,20 +59,25 @@ public EnvSinglenodeSparkIceberg(Standard standard, Hadoop hadoop, DockerFiles d
+        builder.configureContainer(TESTS, dockerContainer -> {
+            dockerContainer.withCopyFileToContainer(
+                    forHostPath(dockerFiles.getDockerFilesHostPath("conf/tempto/tempto-configuration-for-hive3.yaml")),
+                    CONTAINER_TEMPTO_PROFILE_CONFIG);
```
CONTAINER_TEMPTO_PROFILE_CONFIG is deprecated. what's the current way to accomplish this?

jirassimok replied:
It lists `EnvironmentContainers.configureTempto` as its replacement, but that method doesn't work in this case (because that only works when the tempto configuration has a specific name and is in the environment's config directory).

Quite a few things around the environment configuration need to be refactored, and I avoided changing it as much as possible.

(This is also copied verbatim from `EnvSinglenodeHdp3`, so if we change it here, we should probably also change it there.)

findepi replied:
so CONTAINER_TEMPTO_PROFILE_CONFIG is deprecated but we cannot use the replacement?
was it deprecated too early? cc @kokosing

jirassimok replied:
The replacement works in most cases, but there are a few where it doesn't. To address that, we should probably add an overload like `configureTempto(Environment.Builder, String)` to specify which file to use rather than taking a file from a `ResourceProvider`.

findepi commented on testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/env/environment/EnvSinglenodeSparkIceberg.java:
```diff
@@ -57,20 +59,25 @@ public EnvSinglenodeSparkIceberg(Standard standard, Hadoop hadoop, DockerFiles d
     @Override
     public void extendEnvironment(Environment.Builder builder)
     {
+        String dockerImageName = "ghcr.io/trinodb/testing/hdp3.1-hive:" + hadoopImagesVersion;
```
looks redundant, perhaps inline it back.

jirassimok replied:
This came from `EnvSinglenodeHdp3`. This first parts of the setup here are copied verbatim from there, and I separated it to make it visually similar.

I think one unnecessary variable declaration is fine in exchange for making the code duplication more visible (and hopefully, one day we'll get rid of the duplication entirely and inline the variable in exactly one place).

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestCreateDropSchema.java:
```diff
@@ -0,0 +1,145 @@
+        onTrino().executeQuery(format("CREATE SCHEMA %s WITH (location = '%s')", name, location));
+    }
+
+    private static String schemaName(String name)
```
inline, to make it clear schema name is randomized in the test

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestCreateDropSchema.java:
```diff
@@ -0,0 +1,145 @@
+        onTrino().executeQuery("USE iceberg.default");
+    }
+
+    private void createSchema(String name)
+    {
+        onTrino().executeQuery(format("CREATE SCHEMA %s", name));
+    }
+
+    private void createSchema(String name, String location)
+    {
+        onTrino().executeQuery(format("CREATE SCHEMA %s WITH (location = '%s')", name, location));
```
I'd inline these, so that's it's clear what we're testing.

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestCreateDropSchema.java:
```diff
@@ -0,0 +1,145 @@
+
+    private void useIceberg()
+    {
+        onTrino().executeQuery("USE iceberg.default");
```
I'd use fully qualified names instead, but whatever

jirassimok replied:
I had that at first, but then if I (or any future developer) forgot to qualify the names anywhere, the tests might pass without actually using Iceberg.

It also makes the variable declarations in the tests a little less nice, because you need the unqualified schema name for the default schema location.

findepi replied:
> 

we should remove default catalog/schema from tempto configuration here
https://github.com/trinodb/trino/blob/4f67c9055626cbd893f97d7cc4ac5d36be399c3f/testing/trino-product-tests/src/main/resources/tempto-configuration.yaml#L34

@jirassimok can you work on that, separately?

findepi commented:
please add an escape hatch like https://github.com/trinodb/trino/pull/10067

jirassimok committed:
Add deleteData parameter to HiveMetastore.dropDatabase

jirassimok committed:
Delete database location after dropping in GlueHiveMetastore

jirassimok committed:
Pass schema type to FileHiveMetastore.getSchemaPath and its callers

jirassimok committed:
Don't put database schema files in the database directories

Instead of /catalog/database/.trinoSchema, the database schemas in
FileHiveMetastore now go in /catalog/.trinoSchema.database.

jirassimok committed:
Rename database object-deletion method in FileHiveMetastore

jirassimok committed:
Fix a typo

jirassimok committed:
Don't delete visibly non-empty directories when dropping Hive schemas

In SemiTransactionalHiveMetastore, check for files before dropping the
schema. Do not request deletion (via HiveMetastore) if files are
visible in the schema location.

A new config property, hive.delete-schema-locations-fallback,
determines the behavior when Trino can't check the file location.
False (the default) will not request deletion in that case.

jirassimok commented:
Updated to be based on #10146. Only the last two commits are part of this PR.

jirassimok commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoCatalogFactory.java:
```diff
@@ -49,7 +51,8 @@ public TrinoCatalogFactory(
             NodeVersion nodeVersion,
-            IcebergSecurityConfig securityConfig)
+            IcebergSecurityConfig securityConfig,
+            HiveConfig hiveConfig)
```
I brought in `HiveConfig` here because it's used specifically to produce a <code>Trino*Hive*Catalog</code>, but maybe I should copy the config property into `IcebergConfig` instead?

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -204,10 +204,16 @@ public synchronized void createDatabase(HiveIdentity identity, Database database
+            metadataFileSystem.mkdirs(databaseMetadataDirectory);
+        }
+        catch (IOException e) {
+            throw new TrinoException(HIVE_METASTORE_ERROR, "Could not write database", e);
```
Could you please add the database information in the exception?

jirassimok replied:
That would go in #10146, but this was just copied from the other "could not write" errors in the class, which also don't give more detailed messages.

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java:
```diff
@@ -215,7 +219,25 @@ public boolean dropNamespace(ConnectorSession session, String namespace)
+                .getLocation()
+                .map(Path::new);
+
+        boolean deleteData = location.map(path -> {
```
There seems to exist the same check routine in two different places:
- `TrinoHiveCatalog.java`
- `SemiTransactionalHiveMetastore.java`

https://github.com/trinodb/trino/pull/9767/files#diff-4437180ac9a56604967e9ae5f4d4aafdc3804d93faa2c0011eca4fbafb83062aR382-R392

Is there any chance to reuse the code?

jirassimok replied:
We could probably put the key shared logic here into a utility method (maybe in `HiveUtil`), something like `locationContainsFiles(Path, Session)`.

Other than that, I don't think there are great options to de-duplicate this code.

findepi replied:
Let's leave it as is for now, and revisit later, if we find a nice way to package that

homar reviewed (approved):
LGTM

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java:
```diff
@@ -215,7 +219,26 @@ public boolean dropNamespace(ConnectorSession session, String namespace)
+                .map(Path::new);
+
+        // If we see files in the schema location, don't delete it.
+        // If we see no files or can't see the location at all, use fallback.
```
> If we see no files

then we delete the directory

> or can't see the location at all, use fallback.

only on this latter case we use fallback

did you mean "when location is not set"?

jirassimok replied:
If we see no files, delete. If we can't see or if `getDatabase` didn't get the location (it always should in practice, even when using default location) use fallback. I'll update the comment.

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestCreateDropSchema.java:
```diff
@@ -0,0 +1,128 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
+
+public class TestCreateDropSchema
```
How should we test this logic with respect to upcoming Iceberg Glue support? (https://github.com/trinodb/trino/pull/10151)
No change requested in this PR, but still worth discussing.
cc @jirassimok @losipiuk @jackye1995 @phd3

findepi reviewed (commented):
two last commits LGTM, rest is https://github.com/trinodb/trino/pull/10146, but not the latest version, so https://github.com/trinodb/trino/pull/10146 needs to be merged and this one rebased.

findepi commented:
merged as ccee7b6af43a8afcda27fe12dcf1c6af81d700e9, thanks
